### ISSUE TITLE
Printf function isn't working

### ISSUE THREAD

**Issue by fjtrujy** [pspdev/pspsdk#56]
During the migration and upgrade of newlib I have faced that `printf` function is not working anymore in the current toolchain.
(@wally4000 double-checked it.)

I have been investigating for some days till I found why it wasn't working.

On the `crt` files we have `__psp_libc_init` as `weak reference`
https://github.com/pspdev/pspsdk/blob/master/src/startup/crt0.c#L41
```c
/* Allow newlib/psplibc to provide an init hook to be called before main */
extern void __psp_libc_init(int argc, char *argv[]) __attribute__((weak));
```

Then in the `_start` method (entrance point of every single execution), we have:
https://github.com/pspdev/pspsdk/blob/master/src/startup/crt0.c#L78
```c
	/* Call libc initialization hook */
	if(__psp_libc_init != NULL)
		__psp_libc_init(argc, argv);
```

Till this point everything looks fine, I think that the developer who implement this, was thinking:
 _"if user links the binary with `libpsplibc.a` then it will call the `__psp_libc_init` function"_

Then on the `build.mak` that we usually use for compilation we have:
https://github.com/pspdev/pspsdk/blob/master/src/base/build.mak#L72
```make
ifeq ($(USE_PSPSDK_LIBC),1)
# Use the pspsdk libc
PSPSDK_LIBC_LIB = -lpsplibc
CFLAGS := -I$(PSPSDK)/include/libc $(CFLAGS)
else
```

Cool! Everything still looks fine, where is the problem then?
The problem is that the `__psp_libc_init` definition in the `libpsplibc.a` is as well being `weak`:
https://github.com/pspdev/pspsdk/blob/master/src/libc/init.c#L20
```c
__attribute__((weak, constructor))
void _pspsdk_libc_init()
{
    _pspsdk_alloc_init();
    _pspsdk_stdio_init();
    _pspsdk_stdlib_init();
}
```

So basically all the references of `__psp_libc_init` in the binary are weak, with the consequences of `__psp_libc_init` never being executed even if the user adds the `libpsplibc.a` during the linking process.

In my opinion, there are 2 solutions:

1.  To remove the weak refecence in the crt files
2.  To remove the weak reference in the libpsplibc.a

I honestly prefer the first one, and it is how I have done in the toolchain that I'm preparing.

Thanks

**Comment by liclac** [comment]
This is some impressive detective work!
Doing that makes sense to me. 👍

**Comment by carstene1ns** [comment]
There is one thing not right here: You either link with newlib (default, more mature) or psplibc (small, not full-featured), but never both.
If you are working on updating newlib, there should be no reason to touch psplibc.

**Comment by fjtrujy** [comment]
This issue can be now closed.
Thanks
